### ISSUE TITLE
Breaking change in vcat between 0.6.2 and 0.6.3

### ISSUE THREAD

**Issue by a-parton** [JuliaLang/julia#27652]
Between 0.6.2 and 0.6.3 an error appears to have been introduced when concatenating an array with an empty array. I've looked through the commits and the only change that appears to affect vcat is an improvement to the support of abstract arrays, which makes me think that this might not have been a deliberate move.

In 0.6.3:
```
julia> versioninfo()
Julia Version 0.6.3
Commit d55cadc350 (2018-05-28 20:20 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i5-7200U CPU @ 2.50GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, broadwell)

julia> vcat([],[1.0 2.0 3.0])
ERROR: ArgumentError: number of columns of each array must match (got (1, 3))
Stacktrace:
 [1] typed_vcat(::Type{Any}, ::Array{Any,1}, ::Array{Float64,2}) at ./abstractarray.jl:1141
 [2] vcat(::Array{Any,1}, ::Array{Float64,2}) at ./sparse/sparsevector.jl:990
```
 
and in 0.6.2:
```
julia> versioninfo()
Julia Version 0.6.2
Commit d386e40c17 (2017-12-13 18:08 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin14.5.0)
  CPU: Intel(R) Xeon(R) CPU E5-1680 v2 @ 3.00GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Sandybridge)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, ivybridge)

julia> vcat([],[1.0 2.0 3.0])
1×3 Array{Any,2}:
 1.0  2.0  3.0
```

**Comment by mbauman** [comment]
Thanks for the report!  This is also the case on 0.7, and I think this was indeed an unintended break in behavior for the backport.  A quick bisect confirms that 72c3e62ba2fbab5a7572269abe706217eb93c7e2 was indeed the culprit, intending to just fix `vcat(1:8, fill(1, (4, 1)))`.

Empty arrays still work, but they must be `0x3`:

```julia
julia> vcat(reshape([],0,3),[1.0 2.0 3.0])
1×3 Array{Any,2}:
 1.0  2.0  3.0
```

Should we figure out a way to restore the old behavior on 0.6.4?  We're also a little inconsistent in applying this rule:

```julia
julia> vcat(zeros(0), rand(2,2,2))
2×2×2 Array{Float64,3}:
# …

julia> vcat(zeros(0,0), rand(2,2,2))
ERROR: DimensionMismatch("mismatch in dimension 2 (expected 0 got 2)")
Stacktrace:
 [1] _cs at ./abstractarray.jl:1329 [inlined]
 [2] _cshp at ./abstractarray.jl:1319 [inlined]
 [3] _cshp at ./abstractarray.jl:1326 [inlined]
 [4] cat_shape at ./abstractarray.jl:1304 [inlined] (repeats 2 times)
 [5] _cat_t at ./abstractarray.jl:1340 [inlined]

**Comment by ararslan** [comment]
I've pushed a reversion for that change in #27621.

**Comment by JeffBezanson** [comment]
This has been fixed in a 0.6 point release. In 1.0 we give the error, but it is arguably correct. In other cases I believe we'd need to give more errors, which would be a breaking change. That can be discussed separately.
